pep-0639.rst

```diff
@@ -1,543 +1,2220 @@
+  will be referred to as **Wheel** in Title Case.
+
+
+Specification
+=============
+
+The changes necessary to implement the improved license handling outlined in
+this PEP include those in both
+`distribution package metadata <Core metadata_>`_, as defined in the
+`core metadata specification <#coremetadataspec_>`_, and
+`author-provided project source metadata <Project source metadata_>`_, as
+originally defined in PEP 621.
+
+Further, `minor additions <License files in project formats_>`_ to the
+source distribution (sdist), built distribution (wheel) and installed project
+specifications will help document and clarify the already allowed,
+now formally standardized behavior in these respects.
+Finally, `guidance is established <Converting legacy metadata_>`_
+for tools handling and converting legacy license metadata to license
+expressions, to ensure the results are consistent, correct and unambiguous.
+
+Note that the guidance on errors and warnings is for tools' default behavior;
+they MAY operate more strictly if users explicitly configure them to do so,
+such as by a CLI flag or a configuration option.
+
+
+Core metadata
+-------------
+
+The `PyPA Core Metadata specification <#coremetadataspec_>`_ defines the names
+and semantics of each of the supported fields in the distribution metadata of
+Python distribution packages and installed projects.
+
+This PEP `adds <Add License-Expression field_>`_ the
+``License-Expression`` field,
+`adds <Add License-File field_>`_ the ``License-File`` field,
+`deprecates <Deprecate License field_>`_ the ``License`` field,
+and `deprecates <Deprecate license classifiers_>`_ the license classifiers
+in the ``Classifier`` field.
+
+The error and warning guidance in this section applies to build and
+publishing tools; end-user-facing install tools MAY be more lenient than
+mentioned here when encountering malformed metadata
+that does not conform to this specification.
+
+As it adds new fields, this PEP updates the core metadata to version 2.3.
+
+
+Add ``License-Expression`` field
+''''''''''''''''''''''''''''''''
+
+The ``License-Expression`` optional field is specified to contain a text string
+that is a valid SPDX license expression, as defined herein.
+
+Publishing tools SHOULD issue an informational warning if this field is
+missing, and MAY raise an error. Build tools MAY issue a similar warning,
+but MUST NOT raise an error.
 
-Over time, recommending the usage of these expressions will help Python package
-publishers improve the clarity of their license documentation to the benefit of
-package authors, consumers and redistributors.
+A license expression is a string using the SPDX license expression syntax as
+documented in the `SPDX specification <#spdxpression_>`_, either
+Version 2.2 or a later compatible version.
 
+When used in the ``License-Expression`` field and as a specialization of
+the SPDX license expression definition, a license expression can use the
+following license identifiers:
```

Comment:
Aside from `LicenseRef-Proprietary` there is no real way to specific a custom (proprietary) license type. That isn't an issue for most public packages, but since Python is also used in corporate environments with custom package indexes, I could imaging they might want some more flexibility. Especially as we deprecate the `license` field.

Does it make sense to add something like `LicenseRef-Custom:<some custom license name>`?

Reply:
Hmm, good point. SPDX already specifies the use of `LicenseRef-<CUSTOM-LICENSE-NAME>`, which is what we are taking advantage of here, so if we did allow other custom identifiers, they would presumably want to follow that pattern rather than making up something bespoke. For niche cases that need very specific, proprietary custom licenses, making up an arbitrary bespoke license identifier is not terribly useful, I'd think, since people and tools won't know recognize it unless they are already aware of it anyway, and would need to reference the `License-File`s anyway for the actual terms, which `License Ref-Proprietary` makes explicit. If they do have their own custom private infra, that can still allow it, and they can also still use it in their individual files and in other standardized SPDX interchange formats. I'm just not sure it adds value allowing arbitrary license IDs for all Python packages given the room it creates for ambiguity and mistakes relative to the seemingly small benefit for such specific cases.

We could add an explicit `LicenseRef-Custom` for theoretical cases where a license is FSF- or OSI-approved and thus not technically `LicenseRef-Proprietary` but doesn't yet have a SPDX identifier, but as these sources are generally kept closely in sync (I've been a small part of that), this would only be true for a tiny number of edge cases packages, as far as I'm aware. There could also hypothetically be cases where the author has a strong good faith belief that their bespoke license is not proprietary, but it hasn't (yet) been reviewed by FSF or OSI...but not sure if that does more harm than good to allow specifying that, given the increased ambiguity. And, in any case, in terms of actively maintained modern packages it seems rather unlikely this is more than a handful of cases. As such, my inclination would be to stick with @pombredanne 's original language here (which this PR essentially leaves unchanged) unless there is a clear consensus otherwise.

Reply:
Didn't though of `LicenseRef-<CUSTOM-LICENSE-NAME>`. You are right that `LicenseRef-Proprietary` should cover it. I'm just wondering though if we should at least allow the former for **build** tools. (Maybe it would even make sense if only PyPI would reject them.)

```
When processing the ``License-Expression`` field to determine if it contains
 a valid license expression, build and publishing tools:
 - SHOULD halt execution and raise an error if:
   - The field does not contain a valid license expression
   - One or more license identifiers are not valid (as defined above)   # <-- !
```

--
I don't plan on using this personally, it's just something I though about that could be an issue if the `license` field is eventually removed. This could lower the barrier in this regard.

Reply:
Hmm, that's worth thinking about. Currently, given its a `SHOULD`, custom build and publishing tools and those oriented toward this sort of use case are allowed to not raise an error if the license identifier is unknown, if they have reason to do so (as they would, if there was a need for this). We could carve out an exception "...unless they begin with `LicenseRef-`, in which case build tools SHOULD issue a warning instead.".

However, that does complicate the spec a bit, and means that build tools won't catch misspellings of `LicenseRef-Proprietary` or `LicenseRef-Public-Domain`, or users who prepend `LicenseRef` to their SPDX identifiers (as some mistakenly do), or users who go for a custom license without carefully considering the implications...unless users see the warning, which at present could be quite buried within the very verbose build output and be very difficult to spot, even for Python core devs, a point made on [the Discourse thread](https://discuss.python.org/t/pep-639-improving-license-clarity-with-better-package-metadata/2154/71?u=cam-gerlach). At least anecdotally, this seems more likely than the number of authors who genuinely need to make up their own bespoke license identifiers and use them in the packaging metadata with open source tools.

Since custom and enterprise-oriented build tooling is already free to allow such, which seems likely to be the case if custom internal-only IDs are actually going to be useful for other tools to consume, and the other ways that custom IDs can still be used (SPDX manifest, individual file SPDX tags, REUSE project standard, DEP5 standard, etc, plus actual license file detection, which this PEP makes much easier), on top of the fact that such organizations would have to be using the `License` field for that very specific purpose already for it to be a regression, I'm still not sure its necessary in this PEP.

To note, when NPM implemented SPDX identifiers, they initially had a (very poorly named) `UNLICENSED` identifier and one that was `See license in FILE_PATH` (akin to our `License-File`),, otherwise it didn't allow any `LicenseRef` custom ones at all, including `LicenseRef-Proprietary` and `LicenseRef-Public-Domain`, nor allow a package to not have a license expression unless `private: true` was manually set. They later removed even that, and only then did users complain in droves, after which they later re-implemented `UNLICENSED` following a lot of complaints (see discussion in npm/npm#8918 ). So we'd still be offering clearer and more diverse tags than they do, and there didn't appear to be demand for custom `LicenseRef`s as opposed to what this PEP has, `LicenseRef-Proprietary`. 

My feeling is given the deprecation process will presumably last several years, if there are actual necessary use cases where specifying a custom license more specific than `License Ref-Proprietary` in the `License-Expression` field packaging metadata, using open source tools, for proprietary projects, currently using `license` for this, and for which other options aren't sufficient, we'll hear about it at some point before `license` is actually removed, and we can always loosen the spec later but can't tighten it without breaking backward compat.

Reply:
Some good points you mentioned. Especially "we can change it later if necessary" is quite convincing.

Thanks for taking the time to discuss these issues. I appreciate that!

Reply:
Thanks, same to you! Some great points (and great questions) as well, and it certainly got me thinking. I could still see this happening if there is a clear use case and concrete demand for it.

Pondering it some more, I actually do think your other point, about deferring validation of this to publishing tools, is still something worth considering, since it means that Setuptools and other build tools have to vendor `license-expressions` as a dep (or add their own validation), versus just doing it in `twine` on `twine check` and `twine upload`. It would be good to ensure the metadata is valid earlier than the final upload, or not at all for projects that are just hosted on GitHub or elsewhere or use other upload mechanisms, and build tools will need it anyway if they include publishing functionality, want to do anything in the conversion section or otherwise interact with license expressions, the library to do so is designed explicitly for this purpose, well maintained and lightweight, and I'd be willing to help integrate, but I still think it would be good to hear from tool maintainers about whether this would be pose blocking issues for them (of course, it is still a should, so tools could opt out with a well-motivated reason).

I thought about pinging the main packaging tool maintainers, `@sdispater`,  `@takluyver` and `@frostming`, plus maybe `@bhrutledge` from Twine (under the presumption that we have enough Setuptools-involved people here already, given based on some of the issues that motivated this PEP, licensing issues aren't something `@jaraco` has much time to focus on given everything else on his plate), but I wasn't sure of the etiquette involved. Thoughts?

Reply:
Not sure if pinging any of them here provides much benefit. We will just create another place of discussion that isn't really suited for it. In case you still what to, I would suggest to create a new discourse thread specifically for it.

**As for your question: Should build tools have to validate the `license-expression`?**
Ultimately, I think it would be nice if most (all) of them implement it. From my experience, Python package are uploaded to a surprising number of places not just PyPI. So implementing it only in `twine` would mean that we could miss at least some.

From a maintenance perspective, I suspect this whole PEP will require a lot time. Not just adding the support for the new keys / fields and validation, but users still need to start using a new version for their built tools and add the keys to their config, too. There is also the issue that the initial support might be incomplete. For tools that want to evaluate the `license-expression` I would recommend to implement their own validation.

From that view, it doesn't really matter if it's `SHOULD` or `CAN`. Tools are all likely to skip the validation, at least initially.

Reply:
Okay, I see. As mentioned, I can certainly help with implementation and there is @pombredanne 's existing library specifically as a reference implementation of this PEP, but I agree that it will still take time. What seems to be critical, at least per the comments of several packaging luminaries in the thread and I'd tend to agree, is ensuring that there is at least validation on PyPI's end at or before when initial Metadata v2.3 support is rolled out in production versions of build tools, so we can at least ensure that all packages uploaded to PyPI that declare v2.3 support and define one or more of the new fields are valid from the start. Particularly since we'll presumably need some way to actually display this on PyPI, once the `license` field and license tags are deprecated...

Reply:
Indeed, we could; the guidance as it stands now is based on yours, and I didn't really touch that section. I just worry about catching unintentional mispellings of the two canonically defined ones, or users thinking they have to prepend `LicenseRef` in front of license identifiers, as there unfortunately still seems to be some confusion over. That would also create the need for more cross-field validation, and if `license-files.globs` was used or the default was relied upon, there would be no guarantee that the package would actually contain the named license files unless tools checked the source tree/sdist contents. We'd also have to consider how we'd want PyPI to handle that. I'm not sure the relatively niche use cases where something more specific than `LicenseRef-Proprietary` was needed outweighs that, but I'd ultimately trust your judgement, assuming others are okay with requiring and cross-validating that at least one `License-File` is matched (and additionally, the package may just have a `AUTHORS.md` which it would match, which would not really be a license, so its not necessarily reliable even then).

Reply:
@pombredanne As I understand it, the main reason against allowing `LicenseRef-xyz` is that it will make implementation and validation of License-Expressions much more difficult.

Maybe it's worth to consider a middle ground here. I think we agree that theoretically there might be instances where `LicenseRef-Proprietary`, `LicenseRef-Public-Domain`, and the other valid SPDX identifiers aren't enough. What do you guys think about allowing `LicenseRef-Custom` for these cases?

I would not recommend to enforce that a `License-File` must be provided, though. It would just get too complicated. For instance, what do we do if someone chooses a different name? Instead, anyone who encounters `LicenseRef-Custom` should assume it's equivalent to `LicenseRef-Proprietary` unless stated otherwise in a `License-File`. That's consistent with how it would have been handled anyway if the `License-Expression` was missing.

Reply:
I was actually initially attracted to the `LicenseRef-Custom` approach as well, having initially suggested something like that. However, I'm not sure how much it really adds over `LicenseRef-Proprietary`, since AFAIK (correct me if I'm wrong, @pombredanne ) all of the (FSF-recognized) Free and (OSI-recognized) Open Source licenses (i.e., non-proprietary) have SPDX identifiers. Thus, other licenses would thus implicitly qualify as both non-free and non-open, thus are proprietary, aside from the potential for extremely niche cases of licenses that appear to meet the OSS definition and/or the Free Software definition, but are not yet officially recognized by those organizations, _and_ don't otherwise already have a SPDX identifier. So I'm not really sure if there's a clear use case for `LicenseRef-Custom` that wouldn't be covered by a license expression including `LicenseRef-Proprietary`, `LicenseRef-Public-Domain` and the allowed license identifiers, whereas while allowing custom LicenseRefs comes with more downsides, it also has more practical benefits. If there's interest, though, we could think about allowing them in build tools, warning in publishing tools and not allowing on PyPI?

I also thought about proposing distinguishing `All Rights Reserved` from more permissive but still proprietary licensing, but if a package was really explicitly licensed All Rights Reserved (which is essentially equivalent to not having a license at all), technically speaking outside of any rights granted by PyPI's ToS, no one would be allowed to download or use it, so there doesn't seem to be a clear motivation to package and share it in the first place. But perhaps `LicenseRef-All-Rights-Reserved` could be allowed by packaging tools, but not by publishing tools/PyPI?

Reply:
@pombredanne @cdce8p Regardless of what we decide, I think the outcome of our fruitful and insightful discussion so far should be reflected in the PEP. Should I add it as an open issue, or should I document the current approach more completely in the "rejected ideas" section, leaving the door open for further revision on that point?

Reply:
As mentioned, for now, I added an Open Issue that summarized the points made thus far in our thread above, so they can be included in the PEP and can serve as a basis for further discussion on the Discourse thread, as needed.